paddle/fluid/compiler/piano/note_builder.h

```diff
@@ -0,0 +1,124 @@
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include "glog/logging.h"
+#include "paddle/fluid/compiler/piano/note/opcode.h"
+#include "paddle/fluid/compiler/piano/shape.h"
+#include "paddle/fluid/platform/enforce.h"
+#include "paddle/fluid/platform/errors.h"
+
+namespace paddle {
+namespace piano {
+
+class NoteBuilder;
+
+// A Operand is generally constructed by a NoteBuilder, as the returned value of
+// an Instruction, and can be used to an operand for succeeding instructions
+class Operand {
+ public:
+  using ShapeClass = paddle::piano::Shape;
+
+  Operand() : instr_id_(-1), builder_(nullptr) {}
+
+  // whether this operand valid
+  bool Valid() const { return instr_id_ >= 0 && builder_ != nullptr; }
+
+  // the builder that contructs this operand
+  NoteBuilder* Builder() const {
+    PADDLE_ENFORCE_NE(builder_, nullptr,
+                      platform::errors::InvalidArgument("Builder is nullptr"));
+    return builder_;
+  }
+
+  // shape of this operand
+  const ShapeClass& Shape();
+
+ private:
+  // declare the folloing methods as private and only can be used from friend
+  // class, to prevent from illegal usage in complie-time
+  explicit Operand(NoteBuilder* builder) : instr_id_(-1), builder_(builder) {}
+  Operand(int64_t id, NoteBuilder* builder)
+      : instr_id_(id), builder_(builder) {}
+
+  int64_t Id() const { return instr_id_; }
+
+  friend class NoteBuilder;
+
+ private:
+  // the unique id that denotes which instruction generate this value
+  int64_t instr_id_;
+  // the builder that holds the instruction
+  NoteBuilder* builder_;
+};
+
+// A NoteBuilder keeps a list of instructions within the same Note function,
+// and user can append new instructions with one or more operands which come
+// from instructions enqueued
+//
+// This is used as a convenient interface for building up the initial Note
+// Module.
+class NoteBuilder {
+ public:
+  explicit NoteBuilder(const std::string& name) : name_(name) {}
+
+  // Append an new instruction
+  Operand AppendInstruction(note::InstructionProto&& instr, note::OpCode opcode,
+                            const std::vector<Operand>& operands);
+
+  // Returns the shape of the given operand.
+  const Shape& GetShape(Operand op) const;
+
+  // Build the init note::ModuleProto with an entry function
+  // which includes all instructions
+  note::ModuleProto Build();
+
+  // name of this builder
+  const std::string& name() { return name_; }
+
+ private:
+  // Generate the next sequential id
+  int64_t GetNextId() { return ++next_id_; }
+
+  // Build the signature of entry function
+  Signature BuildSignature() const;
+
+ private:
+  // Name to use for the built note::ModuleProto
+  std::string name_;
+
+  // The next sequential ID for every instruction contained within this builer.
+  int64_t next_id_ = 0;
+
+  // The instructions list
+  std::vector<note::InstructionProto> instructions_;
```

Comment:
Can I assume this instruction list is appending only? Should we remove or add instructions as a result of optimization we don't actually work in the middle of the list. If its a design choice then this is fine. Otherwise vector<Instruction*> is preferred.

Reply:
Yes, we just append instructions here. Instructions and functions will remove/add in later note ir pass.